src/_pytest/python.py

```diff
@@ -911,6 +911,124 @@ def hasnew(obj: object) -> bool:
     return False
 
 
+@final
+@attr.s(frozen=True, auto_attribs=True)
+class IdMaker:
+    argnames: Iterable[str] = attr.Factory(list)
+    parametersets: Iterable[ParameterSet] = attr.Factory(list)
+    idfn: Optional[Callable[[Any], Optional[object]]] = None
+    ids: Optional[List[Union[None, str]]] = None
+    config: Optional[Config] = None
+    nodeid: Optional[str] = None
+
+    def make_unique_test_ids(self) -> List[str]:
+        resolved_ids = list(self._make_resolved_ids())
+
+        # All IDs must be unique!
+        unique_ids = set(resolved_ids)
+        if len(unique_ids) != len(resolved_ids):
+
+            # Record the number of occurrences of each test ID.
+            test_id_counts = Counter(resolved_ids)
+
+            # Map the test ID to its next suffix.
+            test_id_suffixes: Dict[str, int] = defaultdict(int)
+
+            # Suffix non-unique IDs to make them unique.
+            for index, test_id in enumerate(resolved_ids):
+                if test_id_counts[test_id] > 1:
+                    resolved_ids[index] = f"{test_id}{test_id_suffixes[test_id]}"
+                    test_id_suffixes[test_id] += 1
+
+        return resolved_ids
+
+    def _make_resolved_ids(self) -> Iterable[str]:
+        """Combine multiple parameter IDs to a single string for each parameter set."""
```

Comment:
Not entirely accurate. Suggestion: "Resolve IDs for all parameter sets (may contain duplicates)."